pkg/controller/hvpa/hvpa_controller.go

```diff
@@ -0,0 +1,651 @@
+	} else {
+		desiredReplicas = hpaStatus.DesiredReplicas
+	}
+
+	var vpaWeight autoscalingv1alpha1.VpaWeight
+	// lastFraction is set to default 1 to handle the case when vpaWeight is 1 in the matching interval,
+	// and there are no fractional vpaWeights in the previous intervals. So we need to default to this value
+	lastFraction := autoscalingv1alpha1.VpaWeight(1)
+	lookupNextFraction := false
+
+	for _, interval := range hvpa.Spec.WeightBasedScalingIntervals {
+		if lookupNextFraction {
+			if interval.VpaWeight < 1 {
+				vpaWeight = interval.VpaWeight
+				break
+			} else {
+				continue
+			}
+		}
+		// TODO: Following 2 if checks need to be done as part of verification process
+		if interval.StartReplicaCount == 0 {
+			interval.StartReplicaCount = *hvpa.Spec.HpaTemplate.MinReplicas
+		}
+		if interval.LastReplicaCount == 0 {
+			interval.LastReplicaCount = hvpa.Spec.HpaTemplate.MaxReplicas
+		}
+		if interval.VpaWeight < 1 {
+			lastFraction = interval.VpaWeight
+		}
+		if currentReplicas >= interval.StartReplicaCount && currentReplicas <= interval.LastReplicaCount {
+			vpaWeight = interval.VpaWeight
+			if vpaWeight == 1 {
+				if desiredReplicas < currentReplicas {
+					// If HPA wants to scale in, use last seen fractional value as vpaWeight
+					// If there is no such value, we cannot scale in anyway, so keep it default 1
+					vpaWeight = lastFraction
+				} else if desiredReplicas > currentReplicas {
+					// If HPA wants to scale out, use next fractional value as vpaWeight
+					// If there is no such value, we can not scale out anyway, so we will end up with vpaWeight = 1
+					lookupNextFraction = true
+					continue
+				}
+			}
+			break
+		}
+	}
+
+	// Memory for newDeploy is assigned in the function getWeightedRequests
+	newDeploy, resourceChanged, err := getWeightedRequests(vpaStatus, hvpa, vpaWeight, deployment)
+	if err != nil {
+		log.Error(err, "Error in getting weight based requests in new deployment")
+	}
+
+	weightedReplicas, err := getWeightedReplicas(hpaStatus, hvpa, deployment, 1-vpaWeight)
+	if err != nil {
+		log.Error(err, "Error in getting weight based replicas")
+	}
+
+	if weightedReplicas == 0 {
+		weightedReplicas = currentReplicas
+	}
+
+	if currentReplicas != weightedReplicas {
+		if newDeploy == nil {
+			newDeploy = deployment.DeepCopy()
+		}
+		newDeploy.Spec.Replicas = &weightedReplicas
+
+		log.Info("HPA", "Scale horizontally from", currentReplicas, "to", weightedReplicas)
+	}
+	if newDeploy == nil {
+		log.Info("Scaling not required")
+		return 0, false, vpaWeight, nil
+	}
+	log.Info("Scaling required")
+	return weightedReplicas - currentReplicas, resourceChanged, vpaWeight, r.Update(context.TODO(), newDeploy)
+}
+
+func getWeightedReplicas(hpaStatus *autoscaling.HorizontalPodAutoscalerStatus, hvpa *autoscalingv1alpha1.Hvpa, deployment *appsv1.Deployment, hpaWeight autoscalingv1alpha1.VpaWeight) (int32, error) {
+	anno := hvpa.GetAnnotations()
+	if val, ok := anno["hpa-controller"]; ok {
```

Comment:
See the comments [here](https://github.com/ggaurav10/hvpa-controller/pull/5/files#diff-f181d9d7ccc73f657c9c60ab37a9bfd1R4).